packages/relay/src/TransactionManager.ts

```diff
@@ -117,50 +121,76 @@ data         | 0x${transaction.data.toString('hex')}
-    const releaseMutex = await this.nonceMutex.acquire()
-    if (isSameAddress(destination, constants.ZERO_ADDRESS)) {
-      const msg = `Preventing to send transaction with action id ${serverAction} to address(0)! Validate your configuration!`
+  async sendTransaction (txDetails: SendTransactionDetails): Promise<SignedTransactionDetails> {
+    const encodedCall = txDetails.method?.encodeABI() ?? '0x'
+    const maxFeePerGas = parseInt(txDetails.maxFeePerGas ?? await this.gasPriceFetcher.getGasPrice())
+    const maxPriorityFeePerGas = parseInt(txDetails.maxPriorityFeePerGas ?? maxFeePerGas.toString())
+
+    const txCost = toBN(maxFeePerGas).mul(toBN(txDetails.gasLimit))
+    const signerBalance = await this.contractInteractor.getBalance(txDetails.signer)
+    if (txCost.gt(toBN(signerBalance))) {
+      throw new Error(`signer ${txDetails.signer} balance ${signerBalance} too low: tx cost is ${txCost.toString()}`)
+    }
+    if (isSameAddress(txDetails.destination, constants.ZERO_ADDRESS)) {
+      const msg = `Preventing to send transaction with action id ${txDetails.serverAction} to address(0)! Validate your configuration!`
       this.logger.error(msg)
       throw new Error(msg)
     }
+    const releaseMutex = await this.nonceMutex.acquire()
     let signedTransaction: SignedTransaction
     let storedTx: StoredTransaction
     try {
-      const nonce = await this.pollNonce(signer)
-      const txToSign = new Transaction({
-        to: destination,
-        value: value,
-        gasLimit,
-        gasPrice: _gasPrice,
-        data: Buffer.from(encodedCall.slice(2), 'hex'),
-        nonce
-      }, this.rawTxOptions)
+      const nonce = await this.pollNonce(txDetails.signer)
+
+      let txToSign: TypedTransaction
+      if (this.transactionType === TransactionType.TYPE_TWO) {
+        // todo add better maxFeePerGas, maxPriorityFeePerGas defaults
+        txToSign = new FeeMarketEIP1559Transaction({
+          to: txDetails.destination,
+          value: txDetails.value,
+          gasLimit: txDetails.gasLimit,
+          maxFeePerGas,
+          maxPriorityFeePerGas: maxPriorityFeePerGas,
+          data: Buffer.from(encodedCall.slice(2), 'hex'),
+          nonce
+        }, this.rawTxOptions)
+      } else {
+        txToSign = new Transaction({
+          to: txDetails.destination,
+          value: txDetails.value,
+          gasLimit: txDetails.gasLimit,
+          gasPrice: maxFeePerGas,
+          data: Buffer.from(encodedCall.slice(2), 'hex'),
+          nonce
+        }, this.rawTxOptions)
+      }
       // TODO omg! do not do this!
-      const keyManager = this.managerKeyManager.isSigner(signer) ? this.managerKeyManager : this.workersKeyManager
-      signedTransaction = keyManager.signTransaction(signer, txToSign)
+      const keyManager = this.managerKeyManager.isSigner(txDetails.signer) ? this.managerKeyManager : this.workersKeyManager
```

Comment:
very strange API: instead of passing signer as parameter to the method, you put it as a member in the txDetails...

Reply:
What? I didn't change API here